packages/node-resolve/src/resolveId.js

```diff
@@ -0,0 +1,121 @@
+import fs from 'fs';
+import path from 'path';
+import { promisify } from 'util';
+
+import resolve from 'resolve';
+
+import { getPackageName } from './util';
+
+const resolveImportPath = promisify(resolve);
+const readFile = promisify(fs.readFile);
+
+const pathNotFoundError = (subPath, pkgPath) =>
+  new Error(`Package subpath '${subPath}' is not defined by "exports" in ${pkgPath}`);
+
+function mapSubPath(pkgJsonPath, subPath, value) {
+  if (typeof value === 'string') {
+    // mapping is a string, for example { "./foo": "./dist/foo.js" }
+    return value;
+  }
+
+  if (Array.isArray(value)) {
+    // mapping is an array with fallbacks, for example { "./foo": ["foo:bar", "./dist/foo.js"] }
+    return value.find((v) => v.startsWith('./'));
```

Comment:
@guybedford any advice how to test for valid import specifiers?

Reply:
This works. The other things you can optionally check are that there are no `.`, `..` or `node_modules` path segments. But that level of strictness isn't completely necessary here too since Node.js will cover the full edge cases naturally as a target.

Reply:
I'll keep this as is for now then.